hnn_core/gui/gui.py

```diff
@@ -1254,33 +1253,38 @@ def _get_evoked_widget(name, layout, style, location, data=None,
     return drive, drive_box
 
 
-def _get_tonic_widget(name, layout, style, data=None):
-
-    default_data = {
+def _get_tonic_widget(name, tstop_widget, layout, style, data=None):
+    cell_types = ['L2_basket', 'L2_pyramidal', 'L5_basket', 'L5_pyramidal']
+    default_values = {
         'amplitude': 0,
         't0': 0,
-        'tstop': 0
+        'tstop': tstop_widget.value
     }
+    t0 = default_values['t0']
+    tstop = default_values['tstop']
+    default_data = {cell_type: default_values for cell_type in cell_types}
 
-    cell_types = ['L2_basket', 'L2_pyramidal', 'L5_basket', 'L5_pyramidal']
-
-    if isinstance(data, dict):
-        default_data.update(data)
     kwargs = dict(layout=layout, style=style)
+    if isinstance(data, dict):
+        default_data = _update_nested_dict(default_data, data)
 
     amplitudes = dict()
     for cell_type in cell_types:
-        if cell_type in data:
-            default_data.update(data[cell_type])
+        amplitude = default_data[cell_type]['amplitude']
         amplitudes[cell_type] = BoundedFloatText(
-            value=deepcopy(default_data['amplitude']),
```

Comment:
is the deepcopy no longer necessary because the value is assigned to variable above?

Reply:
I think the difference is the use and update of the default_data dictionary. Previously we were using a small single-level default dict containing the amplitude. And it was updated in-place in the for-loop if data was passed in for that cell-type. Because the values were pointing to the same part of the default_data dict, an update would change the other widget values, thus necessitating the deep copy. 

The refactored code constructs a nested default_data dict cell-type as keys. Because the structure has a unique amplitude for each cell-type that doesn't get updated, there isn't a need for the deep copy.